File: backend/src/models/Platform.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';

// Platform attributes interface
export interface PlatformAttributes {
  id: number;
  name: string;
  slug: string;
  manufacturer?: string;
  type: 'console' | 'handheld' | 'pc' | 'mobile';
  generation?: number;
  release_year?: number;
  discontinuation_year?: number;
  logo_url?: string;
  primary_color?: string;
  created_at?: Date;
  updated_at?: Date;
}

interface PlatformCreationAttributes extends Optional<PlatformAttributes, 'id' | 'manufacturer' | 'generation' | 'release_year' | 'discontinuation_year' | 'logo_url' | 'primary_color' | 'created_at' | 'updated_at'> {}

class Platform extends Model<PlatformAttributes, PlatformCreationAttributes> implements PlatformAttributes {
  public id!: number;
  public name!: string;
  public slug!: string;
  public manufacturer?: string;
  public type!: 'console' | 'handheld' | 'pc' | 'mobile';
  public generation?: number;
  public release_year?: number;
  public discontinuation_year?: number;
  public logo_url?: string;
  public primary_color?: string;
  public readonly created_at!: Date;
  public readonly updated_at!: Date;
}

Platform.init(
  {
    id: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true
    },
    name: {
      type: DataTypes.STRING(100),
      allowNull: false
    },
    slug: {
      type: DataTypes.STRING(100),
      allowNull: false,
      unique: true
    },
    manufacturer: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    type: {
      type: DataTypes.ENUM('console', 'handheld', 'pc', 'mobile'),
      allowNull: false,
      field: 'platform_type'
    },
    generation: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    release_year: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    discontinuation_year: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    logo_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    primary_color: {
      type: DataTypes.STRING(7),
      allowNull: true
    },
    created_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    },
    updated_at: {
      type: DataTypes.DATE,
      defaultValue: DataTypes.NOW
    }
  },
  {
    sequelize,
    tableName: 'platforms',
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  }
);

export default Platform;

File: mobile-app/lib/models/platform.dart
enum PlatformType {
  console,
  handheld,
  pc,
  mobile;

  String get apiValue {
    switch (this) {
      case PlatformType.console:
        return 'console';
      case PlatformType.handheld:
        return 'handheld';
      case PlatformType.pc:
        return 'pc';
      case PlatformType.mobile:
        return 'mobile';
    }
  }

  String get label {
    switch (this) {
      case PlatformType.console:
        return 'Console';
      case PlatformType.handheld:
        return 'Handheld';
      case PlatformType.pc:
        return 'PC';
      case PlatformType.mobile:
        return 'Mobile';
    }
  }

  static PlatformType fromString(String? value) {
    switch (value) {
      case 'console':
        return PlatformType.console;
      case 'handheld':
        return PlatformType.handheld;
      case 'pc':
        return PlatformType.pc;
      case 'mobile':
        return PlatformType.mobile;
      default:
        return PlatformType.console;
    }
  }
}

class GamePlatform {
  final int id;
  final String name;
  final String slug;
  final String? manufacturer;
  final PlatformType type;
  final int? generation;
  final int? releaseYear;
  final int? discontinuationYear;
  final String? logoUrl;
  final String? primaryColor;

  GamePlatform({
    required this.id,
    required this.name,
    required this.slug,
    this.manufacturer,
    required this.type,
    this.generation,
    this.releaseYear,
    this.discontinuationYear,
    this.logoUrl,
    this.primaryColor,
  });

  factory GamePlatform.fromJson(Map<String, dynamic> json) {
    return GamePlatform(
      id: json['id'] as int,
      name: json['name'] as String,
      slug: json['slug'] as String,
      manufacturer: json['manufacturer'] as String?,
      type: PlatformType.fromString(json['platform_type'] as String? ?? json['type'] as String?),
      generation: json['generation'] as int?,
      releaseYear: json['release_year'] as int?,
      discontinuationYear: json['discontinuation_year'] as int?,
      logoUrl: json['logo_url'] as String?,
      primaryColor: json['primary_color'] as String?,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'slug': slug,
        'manufacturer': manufacturer,
        'type': type.apiValue,
        'generation': generation,
        'release_year': releaseYear,
        'discontinuation_year': discontinuationYear,
        'logo_url': logoUrl,
        'primary_color': primaryColor,
      };
}
